Add `QueryParams` which deprecates `HttpParameters`

trustin opened the pull request:
Motivation:

- #1567

Modifications:

- Fork our HTTP headers API into:
  - `QueryParams`
  - `QueryParamGetters`
  - `QueryParamsBase`
  - `QueryParamsBuilder`
- Extract the common logic and API between HTTP API and query params API into:
  - `StringMultimap`
  - `StringMultimapGetters`
- Add support for `QueryParams` to annotated services.
- Update documentation.
- Deprecate `HttpParameters`.
- Replace the usage of Netty `QueryString{Encoder,Decoder}` with:
  - `QueryParamGetter.toQueryString()` and `appendQueryString()`
  - `QueryParams.fromQueryString()`
- Add microbenchmarks for `QueryString{Encoder,Decoder}`.

Result:

- Closes #1567
- Nicer and faster way to decode/encode a query string.

trustin commented:
@anuraaga @minwoox @ikhoon PTAL. Will remove the draft tag once I finish all to-dos, but it'd be nice if you could take a look and tell me this is the right direction. :-)

anuraaga commented:
I agree with the idea of not sharing when they are semantically different. But it's just so so much code... I'm not a fan but I see why we need to.

I guess we do want to redefine the API since indeed they are separate concepts but can we at least share code for the storage implementation? I would be ok with just one extra object allocation to have storage as a separate class, or it might be ok to use a base class.

trustin commented:
@anuraaga Will try. :sweat_smile:

trustin commented:
- Added `StringMultimap` and `StringMultimapBuilder` to extract common logic. 
- Tried to introduce a common interface type for multimap builders to ensure all builder interfaces have a common set of builder methods, but it seemed to overcomplicate things, so gave up. Added some test cases to prevent a future mistake instead. See `StringMultimapDerivedApiConsistencyTest`.

trustin commented:
Now ready for review. PTAL @anuraaga @minwoox @ikhoon

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java:
```diff
@@ -89,61 +70,84 @@
+
+    @Override
+    boolean nameEquals(@Nullable AsciiString a, @Nullable CharSequence b) {
+        return a != null && (a == b || a.contentEqualsIgnoreCase(b));
```
Think we can remove `a == b` now since this should have been released a while ago

https://github.com/netty/netty/pull/9130

anuraaga replied:
Note for other comment - here if `a == null` it always returns `false` even if `b == null`

trustin replied:
Good point - Made `nameEquals` accept nonnull parameters and handled null in the super class.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/StringMultimap.java:
```diff
@@ -0,0 +1,1098 @@
+
+    abstract int hashName(IN_NAME name);
+
+    abstract boolean nameEquals(@Nullable NAME a, @Nullable IN_NAME b);
```
If I'm not mistaken, I think it's possible to handle `null` in this class - then `nameEquals` doesn't need nullable parameters which is simpler for the subclasses.

By the way, I looked into this mainly since I noticed the subclasses handle two `null` arguments differently, headers returns `false` while query params returns `true` right now which was surprising.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryParamsBase.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Override
+    boolean nameEquals(@Nullable String a, @Nullable String b) {
+        return Objects.equals(a, b);
```
Note for other comment - here if both are `null` it returns `true`

trustin commented on core/src/main/java/com/linecorp/armeria/common/StringMultimap.java:
```diff
@@ -0,0 +1,1098 @@
+        final int h = hashName(name);
+        final int i = index(h);
+        Entry e = entries[i];
+        // loop until the first entry was found
```
It came from Netty code, but yeah, I made it clearer.

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,341 @@
+ * a new {@link QueryParams} derived from an existing one:</p>
+ *
+ * <pre>{@code
+ * QueryParams params = QueryParams.of("name1", "value0")
```
Thanks. Also fixed the same mistake in `HttpHeaders`.

trustin commented on core/src/main/java/com/linecorp/armeria/common/HttpHeaderGetters.java:
```diff
@@ -282,28 +306,33 @@
      */
+    @Override
     default Stream<String> valueStream(CharSequence name) {
         return Streams.stream(valueIterator(name));
```
Fixed

trustin commented:
@anuraaga BIG pass please :laughing:

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedValueResolver.java:
```diff
@@ -1179,55 +1194,52 @@ public String toString() {
-                        } else if (p != null) {
-                            parameters.putAll(p);
-                        }
+                        params2 = QueryParams.fromQueryString(body);
```
Question: Just to make it sure that we can represent that form data as query parameters, right?

anuraaga replied:
Ah this reminds me I was thinking about this and forgot to comment. The name `QueryParams` might seem like it's tied to URL bar (I've had similar feeling about netty's names too). Perhaps UrlParams? It would be consistent with the mime type for form data, which is urlencoded.

trustin replied:
Hmm... isn't it just a query string sent as a body? `UrlParams` sounds more like it's part of a URL to me than the term 'query'.

ikhoon replied:
> Hmm... isn't it just a query string sent as a body? UrlParams sounds more like it's part of a URL to me than the term 'query'.

+1, I also think query params could be represented by either `GET` url or `POST` body.

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedValueResolver.java:
```diff
@@ -462,7 +464,7 @@ private static AnnotatedValueResolver ofHttpParameter(String name,
+                .resolver(resolver(ctx -> ctx.queryParams().getAll(name),
                                    () -> "Cannot resolve a value from HTTP parameter: " + name))
                 .build();
     }
```
Don't we need rename this method as well?

minwoox commented:
Found this as well. https://github.com/trustin/armeria/blob/immutable_http_params/site/src/sphinx/server-annotated-service.rst#getting-an-http-parameter

trustin commented on core/src/main/java/com/linecorp/armeria/common/StringMultimapBuilder.java:
```diff
@@ -0,0 +1,512 @@
+
+    /**
+     * Makes the current {@link #delegate()} a new {@link #parent()} and clears the current {@link #delegate()}.
+     * Call this method when you create a new {@link HttpHeaders} derived from the {@link #delegate()},
```
Let me just use 'object'.

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParamGetters.java:
```diff
@@ -0,0 +1,351 @@
+     */
+    default String toQueryString() {
+        final StringBuilder buf = new StringBuilder(
+                IntMath.saturatedAdd(IntMath.saturatedMultiply(size(), 8), 16));
```
I was making sure the initial capacity is at least 16, but I guess I have to use `Math.max()` here.

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParamsBuilder.java:
```diff
@@ -0,0 +1,523 @@
+     * equivalent to
+     * <pre>{@code
+     * for (String value : values) {
+     *     params.add(name, value);
```
Let me rename it to `builder`. `paramBuilder` sounds fine, but `builder` is shorter and I guess users already know it's about `QueryParamsBuilder` (because they are browsing its Javadoc).

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @VisibleForTesting
+    static String decodeComponent(String s, int from, int toExcluded) {
```
Can we benchmark this against JDK's `URLDecoder`? Looking at the implementation in Java 13, I don't see any reason for it to be any slower than this code.

anuraaga replied:
Realized the biggest performance advantage is probably not needing to copy into `substring`. Lot's of code for that 😅  If keeping it, I would probably replace the `ByteBuffer` and `CharBuffer` with `byte[]` and `new String(bytes, int, int, charset)`, `ByteBuffer` should basically always be at least as slow as native array.

anuraaga replied:
By the way I found https://github.com/netty/netty/pull/6781/files which switched from `byte[]` to `ByteBuffer`. I'm not sure why - anyways we can consider reworking it here or doing it in Netty and backporting if it ends up being better.

anuraaga replied:
Sent https://github.com/netty/netty/pull/9886 - seeing noticably better performance with an array

anuraaga replied:
And also noticed... we could probably beat Netty using our `ThreadLocalByteArray` :)

trustin replied:
We could optimize this further given that we're always decoding UTF-8.

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractHttpHeadersBuilder.java:
```diff
@@ -15,598 +15,54 @@
-    private HttpHeadersBase delegate;
-    @Nullable
-    private HttpHeadersBase parent;
+abstract class AbstractHttpHeadersBuilder<SELF extends HttpHeadersBuilder>
```
There were just too many parameters, so ... :sweat_smile:

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @see ByteBufUtil#writeUtf8(ByteBuf, CharSequence, int, int)
+     */
+    private static void encodeUtf8Component(StringBuilder uriBuilder, String s) {
```
It'd be good to compare this to guava's `URLEscapers` or JDK's `URLEncoder`

anuraaga replied:
Looked at decoder, now looking more detail at encoder ;) I guess this also has the main advantage of not generating intermediate strings so we probably want to keep it.

But the Netty implementation is quite slow since it works on `CharSequence` (it probably needs `instanceof` and optimized implementations added).

Can we go ahead and do at least 1) and if not too complicated 2)?

1) Only decode when `needEncoding` character found to reduce garbage and processing

https://github.com/google/guava/blob/master/guava/src/com/google/common/net/PercentEscaper.java#L147

2) Detect non-encoded runs and use `StringBuilder.append(String, int, int)`? It looks like that would be significantly faster

anuraaga replied:
Actually - don't worry about it I started working on it for Netty codebase :)

trustin replied:
:laughing: OK!

anuraaga replied:
Sent https://github.com/netty/netty/pull/9887 for 1)

trustin replied:
Thinking of forking `PercentEscaper` to make it encode the entire query params instead of individual names/values.

anuraaga replied:
Ah yeah since we already have all params when encoding that should be the fastest of them all. I don't think we'd be able to do it in Netty - so if you feel up to it go for it :)

trustin replied:
I tried, but I failed to beat Guava, maybe due to how things interact with the optimizer. I always see 5-10% throughput drop whenever I modify `PercentEscaper.escape(String)` to append to a `StringBuilder` instead of returning:

```java
// Baseline:
for (Entry<String, String> e : params) {
    buf.append(guavaEscaper.escape(e.getKey()))
       .append('=')
       .append(guavaEscaper.escape(e.getValue()))
       .append('&');
}
// This is 5~10% slower:
for (Entry<String, String> e : params) {
    final String key = e.getKey();
    final String value = e.getValue();
    // isSafeOctetsOnly() is basically same with PercentEscaper.encode()
    // except that it returns true or false.
    if (isSafeOctetsOnly(key)) {
        buf.append(key);
    } else { ... }
    buf.append('=');
    if (isSafeOctetsOnly(value)) {
        buf.append(value);
    } else { ... }
    buf.append('&');
} 
```

I tried various things like:

- replace `if (isSafeOctetsOnly(...) { } else { }` with a method like:
  ```java
  static void encode(StringBuilder buf, String s) {
      for (int i = 0; i < slen; i++) {
          final char c = s.charAt(i);
          if (c >= safeOctets.length || !safeOctets[c]) {
              encodeUtf8Component(buf, s);
              return;
          }
      }

      buf.append(s);
  }

  for (Entry<String, String> e : params) {
      encode(buf, e.getKey());
      buf.append('=');
      encode(buf, e.getValue());
      buf.append('&');
  } 
  ```
- removing `else {}` completely

trustin replied:
So, let me replace `QueryStringEncoder` with `UrlEscapers` in this PR. Better performance and fewer code. Maybe we can revisit again later.

trustin replied:
What's also interesting is that the performance is still 5~10% lower when we delegate to `appendQueryString()` in `toQueryString()`. Inlining `appendQueryString()` gives an instant boost. (Note that still, we lose 5~10% again once we pass `StringBuilder` as a parameter somewhere.)

trustin replied:
Seems like escape analysis is acting on. Switching to thread-local reused string builder removes the performance gap, i.e. let me continue optimizing ours.

trustin replied:
Optimization is complete. Will post the result later.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/StringMultimap.java:
```diff
@@ -0,0 +1,1126 @@
+
+    // Extension points
+
+    abstract int hashName(IN_NAME name);
```
Can we add javadoc to these? I know it's not public API, but it's important for understanding the implementation of the other classes.

trustin replied:
Sure. Done :-)

trustin commented:
> Found this as well. https://github.com/trustin/armeria/blob/immutable_http_params/site/src/sphinx/server-annotated-service.rst#getting-an-http-parameter

Getting a query parameter, then?

anuraaga commented:
By the way, for the future are we considering `HttpRequest.getQueryParams()` and `AggregatedHttpRequest.getFormParams()`?

trustin commented:
> By the way, for the future are we considering `HttpRequest.getQueryParams()` and `AggregatedHttpRequest.getFormParams()`?

That's a good suggestion. I'm somewhat unsure about what the API will look like in the long term, when we start to consider more complex inputs like multi-part requests. Would it still be a good idea to add the getters for url-encoded forms?

anuraaga commented:
That's a good point about multi-part, it always makes APIs annoying ;) Anyways just curious but not really related to this PR

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java:
```diff
@@ -0,0 +1,195 @@
+
+        final int len = s.length();
+
+        // Skip the leading question marks.
```
Just wanted to skip the initial question mark automatically, but it seems like we need to leave it up to the caller for less ambiguity.

anuraaga replied:
This is definitely a hard point. I forget, does RequestHeaders.query() return with a question mark or not? If no then we're good as is, otherwise I'd probably strip a single question mark and make clear in javadoc that input needs to have the leading question mark of a query string (or alternatively we need to make sure javadoc says the opposite).

trustin replied:
There's no `RequestHeaders.query()`, but `RequestContext.query()` returns without a leading `?`. Anyways, I updated not to strip a question mark. After all, it seemed too ambiguous. :sweat_smile:

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,341 @@
+
+    /**
+     * Returns a new parameters which is the result from the mutation by the specified {@link Consumer}.
+     * This method is a shortcut of:
```
Fixed all `a shortcut of`

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,341 @@
+     * @param queryString the query string with or without leading question mark ({@code '?'}).
+     * @param maxParams   the max number of parameters to decode. If the {@code queryString} contains
+     *                    more parameters than this value, the extra parameters will not be decoded.
+     * @param semicolonIsSeparator whether to treat a semicolon ({@code ';'}) as a separator as well as
```
Renamed

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractHttpHeadersBuilder.java:
```diff
@@ -15,598 +15,54 @@
-    @Nullable
-    private HttpHeadersBase parent;
+abstract class AbstractHttpHeadersBuilder<SELF extends HttpHeadersBuilder>
+        extends StringMultimapBuilder<CharSequence, AsciiString, HttpHeadersBase, SELF> {
```
Good idea. Fixed.

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,341 @@
+     * @return the decoded {@link QueryParams}. An empty {@link QueryParams} is returned
+     *         if {@code queryString} is {@code null}.
+     */
+    static QueryParams fromQueryString(@Nullable String queryString, boolean semicolonIsSeparator) {
```
Reordered

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java:
```diff
@@ -0,0 +1,195 @@
+                case '&':
+                    if (addParam(s, nameStart, valueStart, i, params)) {
+                        paramsLimit--;
+                        if (paramsLimit == 0) {
```
We already silently ignore extra headers while decoding. How about adding some to-do here and addressing it when there is a real use for it?

trustin commented:
Benchmark result for `QueryStringEncoder`. :thinking: 

```
Benchmark                                    Mode  Cnt       Score      Error  Units
QueryStringEncoderBenchmark.armeriaAscii    thrpt   50  776705.322 ± 3682.972  ops/s
QueryStringEncoderBenchmark.armeriaMixed    thrpt   50  178059.265 ± 6122.498  ops/s
QueryStringEncoderBenchmark.armeriaUnicode  thrpt   50  144391.816 ±  815.577  ops/s
QueryStringEncoderBenchmark.guavaAscii      thrpt   50  794491.768 ± 6085.282  ops/s
QueryStringEncoderBenchmark.guavaMixed      thrpt   50  198082.177 ± 1470.667  ops/s
QueryStringEncoderBenchmark.guavaUnicode    thrpt   50  152200.760 ± 1560.334  ops/s
QueryStringEncoderBenchmark.jdkAscii        thrpt   50  288311.945 ± 1273.239  ops/s
QueryStringEncoderBenchmark.jdkMixed        thrpt   50   70419.943 ±  743.201  ops/s
QueryStringEncoderBenchmark.jdkUnicode      thrpt   50   57152.709 ±  568.473  ops/s
QueryStringEncoderBenchmark.nettyAscii      thrpt   50  447956.029 ± 1125.380  ops/s
QueryStringEncoderBenchmark.nettyMixed      thrpt   50  121497.684 ± 3005.426  ops/s
QueryStringEncoderBenchmark.nettyUnicode    thrpt   50  103916.834 ±  506.128  ops/s
```

I'm not sure why we are not faster than or on par with Guava. Mind taking a look, @anuraaga?

anuraaga commented:
Will check the code too - first glance though is the numbers may show we're on par. The error ranges are pretty large and they do intersect. Would need more forks / iterations to be more sure

trustin commented:
That's the result with 10 forks. :sweat_smile:

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,241 @@
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
```
Wonder if this is still needed, or should be replaced or in addition to Guava

trustin replied:
Added Guava copyright header.

anuraaga commented on benchmarks/src/jmh/java/com/linecorp/armeria/common/QueryStringEncoderBenchmark.java:
```diff
@@ -0,0 +1,163 @@
+            mixed.add("foo", "alpha・ベータ")
+                 .add("bar", "ガンマ・delta")
+                 .add("baz", "nothing_无_east_东_horse_马_wind_风")
+                 .add("qux", "santa_🎅_fire_🔥_smile_😊_present_🎁");
```
I don't think it's a reason to skip guava since such long strings are rare, but you may try adding a string that's more than 64 characters - that's the threshold for escape analysis to allocate arrays on the stack which could reasonably be happening. It'd be interesting if that causes the performance to tank.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,241 @@
+                             StringBuilder buf, QueryParamGetters params) {
+        for (Entry<String, String> e : params) {
+            final String name = e.getKey();
+            if (isSafeOctetsOnly(name)) {
```
Let's extract a method for this common code that encodes a String - who knows it could be affecting an optimization threshold

trustin replied:
Done. Looking cleaner anyway :-)

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,241 @@
+        buf.setLength(buf.length() - 1);
+    }
+
+    private static boolean isSafeOctetsOnly(String s) {
```
Compared to Guava, we seem to lose a lot by not returning an index here and using it in `encodeUtf8Component` to skip checking initial characters again.

trustin replied:
Fixed

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,241 @@
+            if (c < 0x80) {
+                if (!isSafeOctet(c)) {
+                    if (i > safeOctetStart) {
+                        buf.append(s, safeOctetStart, i);
```
This `safeOctetStart` was just an idea, but I don't know if it helps or hurts in practice given the extra branches 😅 It is very much dependent on the content, and while I think for longer strings (which we'd expect from forms) it could help, for now, probably better to keep it simpler for easier comparison to other implementations, and we can revisit later. It's probably worth doing if it doesn't negatively impact numbers too much as I'd expect it to have small negative impact on small input, large positive impact on larger input.

trustin replied:
I don't know. At the moment we lose at decoding long query string although the gap is small. Will investigate a little bit more.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,241 @@
+     */
+    private static void encodeUtf8Component(TemporaryThreadLocals tempThreadLocals,
+                                            StringBuilder buf, String s) {
+        final char[] tmp = tempThreadLocals.charArray(12);
```
Probably better to allocate a normal array, it's small enough that it should be on the stack. Though the escape analysis is hit-or-miss here - if you allocate in each method it's almost guaranteed to work though and is similar to what guava's doing. Also, since this array has dynamic size (as opposed to `new char[12]`), Java can't remove range checks which hurts performance.

trustin replied:
Tried already and it seems like escape analysis does not kick in sadly.

trustin replied:
Escape analysis works as expected after manually inlining `writeUtf8Surrogate()` :tada:

trustin commented:
Oh, OK. According to `QueryStringEncoderBenchmark` it's still like 1.5-2x faster than Netty's. Let me switch to a simpler version with more branches.

trustin commented:
@anuraaga I've just pushed 937779c569835d2a4caa4d7f890f237746fc9ea1, but it seems like it's 10% slower than DFA version. Could you check what could be improved?

trustin commented:
The result from fd1c4e52e560d5ea78f588344bf17b8b4261cf72
```
Benchmark                                    Mode  Cnt       Score       Error  Units
QueryStringDecoderBenchmark.armeriaAscii    thrpt   50  380672.412 ±  8104.627  ops/s
QueryStringDecoderBenchmark.armeriaLong     thrpt   50   17063.315 ±   617.047  ops/s
QueryStringDecoderBenchmark.armeriaMixed    thrpt   50  119307.118 ±  2963.182  ops/s
QueryStringDecoderBenchmark.armeriaUnicode  thrpt   50   85732.344 ±  2011.455  ops/s
QueryStringDecoderBenchmark.nettyAscii      thrpt   50  343379.090 ±  9079.795  ops/s
QueryStringDecoderBenchmark.nettyLong       thrpt   50    9065.173 ±   249.774  ops/s
QueryStringDecoderBenchmark.nettyMixed      thrpt   50   61184.914 ±   354.602  ops/s
QueryStringDecoderBenchmark.nettyUnicode    thrpt   50   49656.054 ±   558.771  ops/s
QueryStringEncoderBenchmark.armeriaAscii    thrpt   50  761883.482 ± 16652.428  ops/s
QueryStringEncoderBenchmark.armeriaLong     thrpt   50   21469.565 ±   500.053  ops/s
QueryStringEncoderBenchmark.armeriaMixed    thrpt   50  210244.395 ±  4789.462  ops/s
QueryStringEncoderBenchmark.armeriaUnicode  thrpt   50  204346.461 ±   938.928  ops/s
QueryStringEncoderBenchmark.guavaAscii      thrpt   50  774993.351 ±  2737.659  ops/s
QueryStringEncoderBenchmark.guavaLong       thrpt   50   22641.740 ±   150.873  ops/s
QueryStringEncoderBenchmark.guavaMixed      thrpt   50  195272.956 ±  1802.613  ops/s
QueryStringEncoderBenchmark.guavaUnicode    thrpt   50  138458.690 ±  1285.093  ops/s
QueryStringEncoderBenchmark.jdkAscii        thrpt   50  264616.790 ±   620.668  ops/s
QueryStringEncoderBenchmark.jdkLong         thrpt   50    7872.475 ±    44.410  ops/s
QueryStringEncoderBenchmark.jdkMixed        thrpt   50   63819.362 ±   303.451  ops/s
QueryStringEncoderBenchmark.jdkUnicode      thrpt   50   51462.477 ±   179.308  ops/s
QueryStringEncoderBenchmark.nettyAscii      thrpt   50  443927.583 ±   783.943  ops/s
QueryStringEncoderBenchmark.nettyLong       thrpt   50   14472.656 ±   163.951  ops/s
QueryStringEncoderBenchmark.nettyMixed      thrpt   50  124317.750 ±  4115.201  ops/s
QueryStringEncoderBenchmark.nettyUnicode    thrpt   50  102374.791 ±   137.208  ops/s
```

trustin commented:
541796b7ee627e0906fa7544c3c9de2645e40d23 helped a lot. 8383ef964a80008464d5c7fa4b669a506d2139ad helped a little bit.

trustin commented:
@anuraaga @minwoox Ready for another round.

trustin commented:
Slightly better numbers after 8383ef964a80008464d5c7fa4b669a506d2139ad. Now we are on par with or better than Guava except for one case (ASCII).
```
Benchmark                                    Mode  Cnt       Score      Error  Units
QueryStringEncoderBenchmark.armeriaAscii    thrpt   50  751692.461 ± 2123.210  ops/s
QueryStringEncoderBenchmark.armeriaLong     thrpt   50   21553.386 ±  839.109  ops/s
QueryStringEncoderBenchmark.armeriaMixed    thrpt   50  221688.950 ± 1553.724  ops/s
QueryStringEncoderBenchmark.armeriaUnicode  thrpt   50  211791.166 ± 1017.269  ops/s
```

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java:
```diff
@@ -0,0 +1,276 @@
+    }
+
+    private static int decodeHexNibble(char c) {
+        return c < OCTETS_TO_HEX.length ? OCTETS_TO_HEX[c] : -1;
```
https://github.com/netty/netty/pull/9896 is coming too :)

We definitely should add a check vs 0 too to try to eliminate JVM's bounds check. I suspect converting to `int` first could be important since the value in the comparisons is automatically converted to `int` and it wouldn't surprise me if `(int) c` for the conditional and array indexing are treated as different values.

trustin replied:
Optimized

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java:
```diff
@@ -0,0 +1,276 @@
+
+    @SuppressWarnings("checkstyle:AvoidEscapedUnicodeCharacters")
+    private static final char UNKNOWN_CHAR = '\uFFFD';
+    private static final byte[] OCTETS_TO_HEX = new byte['f' + 1];
```
I may stick with `int[]` since the difference in size is so minor, but it could give the JVM a better time of finding some optimization.

trustin replied:
I'm not sure. I'd prefer `byte[]` because it's more compact and thus better utilization of CPU cache.

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/QueryStringEncoder.java:
```diff
@@ -0,0 +1,200 @@
+
+    private static StringBuilder encodeComponent(StringBuilder buf, String s) {
+        final int firstUnsafeOctetIdx = indexOfUnsafeOctet(s, 0);
+        if (firstUnsafeOctetIdx < 0) {
```
I don't know if it's worth restructuring the code, but I guess we have one extra branch here vs Guava because we have a separate method for `indexOfUnsafeOctet` while inlining it here removes the branch - could be affecting the ASCII-only case.

trustin replied:
Nice find. Updated.

franz1981 commented:
Beware stack allocations of arrays ie https://mobile.twitter.com/forked_franz/status/1193481376122777601

trustin commented:
Result from: 1cb66cbe433b112baee655c9d5cf43468e7889c1
```
Benchmark                                    Mode  Cnt       Score       Error  Units
QueryStringEncoderBenchmark.armeriaAscii    thrpt    5  796213.603 ± 17510.214  ops/s
QueryStringEncoderBenchmark.armeriaLong     thrpt    5   22438.634 ±   300.509  ops/s
QueryStringEncoderBenchmark.armeriaMixed    thrpt    5  214785.444 ±  5345.319  ops/s
QueryStringEncoderBenchmark.armeriaUnicode  thrpt    5  211279.841 ±  3465.634  ops/s
QueryStringEncoderBenchmark.guavaAscii      thrpt    5  769711.170 ± 12697.817  ops/s
QueryStringEncoderBenchmark.guavaLong       thrpt    5   22709.233 ±   377.663  ops/s
QueryStringEncoderBenchmark.guavaMixed      thrpt    5  197058.689 ±  8895.232  ops/s
QueryStringEncoderBenchmark.guavaUnicode    thrpt    5  139209.275 ±  1154.131  ops/s
```

trustin commented:
@franz1981 wrote:
> Beware stack allocations of arrays ie mobile.twitter.com/forked_franz/status/1193481376122777601

Thanks for the information. Does it mean JVM will not stack-allocate any array whose length is derived from user input? e.g.

```java
static void func(int length) {
    final int values[] = new int[length];
    // ... `values` does not escape ...
}
```

franz1981 commented:
@trustin I need to re-read the heuristic of the EA, but looking at the simple JMH bench I've sent, yes: you are both limited on the array size by a JVM params and it seems OpenJDK (!= GraalVM) is more conservative, so I suggest to run the benchmarks with `-prof gc` or just use a profiler (async-profiler) to check allocations during your bench...
I wouldn't rely on the JVM to stack allocate things given the latest results I've got on this...my 2 cents.

trustin commented:
Thanks a lot, @franz1981 

Complete result from https://github.com/line/armeria/commit/1cb66cbe433b112baee655c9d5cf43468e7889c1 with 10 forks:
```
Benchmark                                    Mode  Cnt       Score      Error  Units
QueryStringDecoderBenchmark.armeriaAscii    thrpt   50  378741.654 ± 6659.789  ops/s
QueryStringDecoderBenchmark.armeriaLong     thrpt   50   17363.007 ±  711.308  ops/s
QueryStringDecoderBenchmark.armeriaMixed    thrpt   50  120362.631 ± 2517.298  ops/s
QueryStringDecoderBenchmark.armeriaUnicode  thrpt   50   86819.473 ± 1580.938  ops/s
QueryStringDecoderBenchmark.nettyAscii      thrpt   50  352954.120 ± 1303.213  ops/s
QueryStringDecoderBenchmark.nettyLong       thrpt   50    9216.124 ±  285.652  ops/s
QueryStringDecoderBenchmark.nettyMixed      thrpt   50   62038.399 ±  453.026  ops/s
QueryStringDecoderBenchmark.nettyUnicode    thrpt   50   49561.678 ±  659.775  ops/s
QueryStringEncoderBenchmark.armeriaAscii    thrpt   50  802149.965 ± 4023.009  ops/s
QueryStringEncoderBenchmark.armeriaLong     thrpt   50   21525.055 ±  801.401  ops/s
QueryStringEncoderBenchmark.armeriaMixed    thrpt   50  224702.833 ± 1475.840  ops/s
QueryStringEncoderBenchmark.armeriaUnicode  thrpt   50  210494.132 ± 8571.742  ops/s
QueryStringEncoderBenchmark.guavaAscii      thrpt   50  774172.127 ± 4041.075  ops/s
QueryStringEncoderBenchmark.guavaLong       thrpt   50   22669.871 ±  184.624  ops/s
QueryStringEncoderBenchmark.guavaMixed      thrpt   50  199361.612 ± 1562.677  ops/s
QueryStringEncoderBenchmark.guavaUnicode    thrpt   50  139650.766 ±  991.154  ops/s
QueryStringEncoderBenchmark.jdkAscii        thrpt   50  265860.618 ±  511.369  ops/s
QueryStringEncoderBenchmark.jdkLong         thrpt   50    7932.388 ±   59.079  ops/s
QueryStringEncoderBenchmark.jdkMixed        thrpt   50   64110.415 ±  299.827  ops/s
QueryStringEncoderBenchmark.jdkUnicode      thrpt   50   51915.861 ±  145.560  ops/s
QueryStringEncoderBenchmark.nettyAscii      thrpt   50  438574.293 ± 8949.848  ops/s
QueryStringEncoderBenchmark.nettyLong       thrpt   50   14510.677 ±  159.799  ops/s
QueryStringEncoderBenchmark.nettyMixed      thrpt   50  126900.454 ± 1168.554  ops/s
QueryStringEncoderBenchmark.nettyUnicode    thrpt   50  101262.449 ± 1587.243  ops/s
```

trustin commented on benchmarks/src/jmh/java/com/linecorp/armeria/common/QueryStringDecoderBenchmark.java:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2018 LINE Corporation
```
:rofl:

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,348 @@
+     */
+    static QueryParams fromQueryString(@Nullable String queryString, int maxParams,
+                                       boolean semicolonAsSeparator) {
+        if (queryString == null || queryString.isEmpty()) {
```
Now using `Strings.isNullOrEmpty()`

trustin commented on core/src/main/java/com/linecorp/armeria/common/QueryParams.java:
```diff
@@ -0,0 +1,348 @@
+     */
+    default QueryParams withMutations(Consumer<QueryParamsBuilder> mutator) {
+        final QueryParamsBuilder builder = toBuilder();
+        mutator.accept(builder);
```
Fixed

trustin commented:
Thanks everyone for reviews and advices. :+1:
